control-plane/pkg/apis/messaging/v1beta1/kafka_channel_validation.go

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+
+	"knative.dev/eventing/pkg/apis/eventing"
+	"knative.dev/pkg/apis"
+)
+
+func (c *KafkaChannel) Validate(ctx context.Context) *apis.FieldError {
+	errs := c.Spec.Validate(ctx).ViaField("spec")
+
+	// Validate annotations
+	if c.Annotations != nil {
+		if scope, ok := c.Annotations[eventing.ScopeAnnotationKey]; ok {
+			if scope != "namespace" && scope != "cluster" {
+				iv := apis.ErrInvalidValue(scope, "")
+				iv.Details = "expected either 'cluster' or 'namespace'"
+				errs = errs.Also(iv.ViaFieldKey("annotations", eventing.ScopeAnnotationKey).ViaField("metadata"))
+			}
+		}
+	}
+
+	return errs
+}
+
+func (cs *KafkaChannelSpec) Validate(ctx context.Context) *apis.FieldError {
+	var errs *apis.FieldError
+
+	if cs.NumPartitions <= 0 {
+		fe := apis.ErrInvalidValue(cs.NumPartitions, "numPartitions")
+		errs = errs.Also(fe)
+	}
+
+	if cs.ReplicationFactor <= 0 {
+		fe := apis.ErrInvalidValue(cs.ReplicationFactor, "replicationFactor")
+		errs = errs.Also(fe)
+	}
+
+	for i, subscriber := range cs.SubscribableSpec.Subscribers {
+		if subscriber.ReplyURI == nil && subscriber.SubscriberURI == nil {
+			fe := apis.ErrMissingField("replyURI", "subscriberURI")
+			fe.Details = "expected at least one of, got none"
+			errs = errs.Also(fe.ViaField(fmt.Sprintf("subscriber[%d]", i)).ViaField("subscribable"))
+		}
```

Comment:
Question: with this check I can have a ReplyURI and no SubscriberURI, how do I get a reply without a subscriber?

Reply:
Hmm, not sure.

Frankly, these API code are simply copy&pastes from the Kafka Channel in eventing-kafka.

I see the very same check is also in IMC.

Can you create a ticket for this? Let's ask @lionelvillard 's opinion on it.

Reply:
We should have an answer in https://github.com/knative/specs/pull/30/files (still open)

Reply:
Let's not block _this_ PR on that decision spec PR...